runtime/lib/capi/ExecutionContext.hpp

```diff
@@ -193,55 +213,70 @@ class ExecutionContext final {
     [[nodiscard]] QuantumDevice *loadDevice(std::string filename)
     {
         _driver_so_ptr = std::make_unique<SharedLibraryManager>(filename);
-        void *f_ptr = _driver_so_ptr->getSymbol("getCustomDevice");
-        return f_ptr ? reinterpret_cast<decltype(getCustomDevice) *>(f_ptr)() : nullptr;
+        std::string factory_name{_device_name + "Factory"};
+        void *f_ptr = _driver_so_ptr->getSymbol(factory_name);
+        return f_ptr ? reinterpret_cast<decltype(GenericDeviceFactory) *>(f_ptr)(_tape_recording,
+                                                                                 _device_kwargs)
+                     : nullptr;
     }
 
-    [[nodiscard]] bool initDevice(std::string_view name)
+    [[nodiscard]] bool initDevice(std::string_view rtd_lib)
     {
-        if (name != "default") {
-            _device_name = name;
-        }
-
-        if (_device_name == "braket.aws.qubit" || _device_name == "braket.local.qubit") {
-            _device_kwargs = "device_type : " + _device_name + "," + _device_kwargs;
-            _device_name = "openqasm";
-        }
-
+        // Reset the driver pointer
         _driver_ptr.reset(nullptr);
 
-        auto iter = _device_map.find(_device_name);
-        if (iter != _device_map.end()) {
-            _driver_ptr = iter->second(_tape_recording, _device_kwargs);
-
-#ifdef __device_openqasm
-            if (_device_name == "openqasm" && !Py_IsInitialized()) {
-                _py_guard =
-                    std::make_unique<Device::OpenQasm::PythonInterpreterGuard>(); // LCOV_EXCL_LINE
+        // LCOV_EXCL_START
+        // TODO: This can be added to a dictionary or function
+        {
+            // The following is required for C++ tests where these
+            // backend devices are linked in the interface library
+            // of the runtime.
+            // Besides, it provides support for libraries added to
+            // the system path.
+            if (rtd_lib == "lightning.qubit") {
+                _device_name = "LightningSimulator";
+#ifdef __linux__
+                rtd_lib = "librtd_lightning.so";
+#elif defined(__APPLE__)
+                rtd_lib = "librtd_lightning.dylib";
+#endif
```

Comment:
Can we have different functions here? Example pseudcode

```c
inline char* get_suffix_apple() {
#ifdef defined(_APPLE__)
  return ".dylib";
#else
  return NULL;
#endif
}

inline char* get_suffix_linux() {
#ifdef __linux__
  return ".so";
#else
  return NULL;
#endif
}

char* get_suffix_platform() {
  return get_suffix_linux() | get_suffix_apple();
}
```

Reply:
Oh yes, that's the motivation behind the added TODO in code comments.

Reply:
Oh yes, that's the motivation behind the added TODO in code comments a few lines above.

Reply:
I still see the TODO (?) Have you pushed? It is a minor point, I just really dislike preprocessor macros in the middle of functions and prefer functions that just compile away to no-ops.

Reply:
There shouldn't be any TODO :thinking: You may want to referesh your browser to see the latest changes in GH.

Me too. I also don't like preprocessors in the middle of functions. In runtime, we don't rely on these pp macros in too many places; they're used in short methods and mostly in `ExecutionContext.hpp`.